File: mobile/lib/core/protocol/schema.dart
import 'dart:typed_data';
import 'package:msgpack_dart/msgpack_dart.dart' as msgpack;

/// Action Types supported by OpenDeck
enum ActionType {
  hotkey,
  shell,
  media,
  obsAction,
  multiAction,
}

extension ActionTypeExtension on ActionType {
  String toValue() {
    switch (this) {
      case ActionType.hotkey:
        return 'HOTKEY';
      case ActionType.shell:
        return 'SHELL';
      case ActionType.media:
        return 'MEDIA';
      case ActionType.obsAction:
        return 'OBS_ACTION';
      case ActionType.multiAction:
        return 'MULTI_ACTION';
    }
  }

  static ActionType fromValue(String value) {
    switch (value) {
      case 'HOTKEY':
        return ActionType.hotkey;
      case 'SHELL':
        return ActionType.shell;
      case 'MEDIA':
        return ActionType.media;
      case 'OBS_ACTION':
        return ActionType.obsAction;
      case 'MULTI_ACTION':
        return ActionType.multiAction;
      default:
        return ActionType.hotkey;
    }
  }
}

/// Command Action Payload (Phone -> Desktop over Command Characteristic)
class ActionPayload {
  final String id;
  final ActionType actionType;
  final List<String> modifiers;
  final String key;
  final String payload;
  final int sequenceDelayMs;

  ActionPayload({
    required this.id,
    required this.actionType,
    required this.modifiers,
    required this.key,
    required this.payload,
    required this.sequenceDelayMs,
  });

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'actionType': actionType.toValue(),
      'modifiers': modifiers,
      'key': key,
      'payload': payload,
      'sequenceDelayMs': sequenceDelayMs,
    };
  }

  factory ActionPayload.fromMap(Map<String, dynamic> map) {
    return ActionPayload(
      id: map['id'] as String,
      actionType: ActionTypeExtension.fromValue(map['actionType'] as String),
      modifiers: List<String>.from(map['modifiers'] ?? []),
      key: map['key'] as String,
      payload: map['payload'] as String,
      sequenceDelayMs: map['sequenceDelayMs'] as int,
    );
  }

  Uint8List serialize() {
    return msgpack.serialize(toMap());
  }

  factory ActionPayload.deserialize(Uint8List bytes) {
    final map = msgpack.deserialize(bytes) as Map<dynamic, dynamic>;
    return ActionPayload.fromMap(Map<String, dynamic>.from(map));
  }
}

/// System Metrics snapshot
class SystemMetrics {
  final double cpu;
  final double ram;
  final bool micMuted;
  final bool audioPlaying;

  SystemMetrics({
    required this.cpu,
    required this.ram,
    required this.micMuted,
    required this.audioPlaying,
  });

  Map<String, dynamic> toMap() {
    return {
      'cpu': cpu,
      'ram': ram,
      'micMuted': micMuted,
      'audioPlaying': audioPlaying,
    };
  }

  factory SystemMetrics.fromMap(Map<String, dynamic> map) {
    return SystemMetrics(
      cpu: (map['cpu'] as num).toDouble(),
      ram: (map['ram'] as num).toDouble(),
      micMuted: map['micMuted'] as bool,
      audioPlaying: map['audioPlaying'] as bool,
    );
  }
}

/// Telemetry Payload (Desktop -> Phone over Telemetry Characteristic)
class TelemetryPayload {
  final String status;
  final String activeApp;
  final SystemMetrics metrics;

  TelemetryPayload({
    required this.status,
    required this.activeApp,
    required this.metrics,
  });

  Map<String, dynamic> toMap() {
    return {
      'status': status,
      'activeApp': activeApp,
      'metrics': metrics.toMap(),
    };
  }

  factory TelemetryPayload.fromMap(Map<String, dynamic> map) {
    return TelemetryPayload(
      status: map['status'] as String,
      activeApp: map['activeApp'] as String,
      metrics: SystemMetrics.fromMap(Map<String, dynamic>.from(map['metrics'] as Map)),
    );
  }

  Uint8List serialize() {
    return msgpack.serialize(toMap());
  }

  factory TelemetryPayload.deserialize(Uint8List bytes) {
    final map = msgpack.deserialize(bytes) as Map<dynamic, dynamic>;
    return TelemetryPayload.fromMap(Map<String, dynamic>.from(map));
  }
}

/// Handshake & Pairing Payload (Bidirectional over Auth Characteristic)
class HandshakePayload {
  final String clientId;
  final String clientPublicKey;
  final String authCode;

  HandshakePayload({
    required this.clientId,
    required this.clientPublicKey,
    required this.authCode,
  });

  Map<String, dynamic> toMap() {
    return {
      'clientId': clientId,
      'clientPublicKey': clientPublicKey,
      'authCode': authCode,
    };
  }

  factory HandshakePayload.fromMap(Map<String, dynamic> map) {
    return HandshakePayload(
      clientId: map['clientId'] as String,
      clientPublicKey: map['clientPublicKey'] as String,
      authCode: map['authCode'] as String,
    );
  }

  Uint8List serialize() {
    return msgpack.serialize(toMap());
  }

  factory HandshakePayload.deserialize(Uint8List bytes) {
    final map = msgpack.deserialize(bytes) as Map<dynamic, dynamic>;
    return HandshakePayload.fromMap(Map<String, dynamic>.from(map));
  }
}

File: desktop/src-tauri/src/ble/schema.rs
use serde::{Deserialize, Serialize};

/// BLE UUID constants for the OpenDeck GATT service and characteristics
pub struct BleUuid;
impl BleUuid {
    pub const SERVICE:   &'static str = "13370001-DEAD-BEEF-FEED-CAFE00000001";
    pub const COMMAND:   &'static str = "13370002-DEAD-BEEF-FEED-CAFE00000001";
    pub const TELEMETRY: &'static str = "13370003-DEAD-BEEF-FEED-CAFE00000001";
    pub const AUTH:      &'static str = "13370004-DEAD-BEEF-FEED-CAFE00000001";
}

/// Action type triggered by phone button tap
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ActionType {
    Hotkey,
    Shell,
    Media,
    ObsAction,
    MultiAction,
}

/// Command Action Payload (Phone -> Desktop over Command Characteristic)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ActionPayload {
    pub id: String,
    pub action_type: ActionType,
    pub modifiers: Vec<String>,
    pub key: String,
    pub payload: String,
    pub sequence_delay_ms: u32,
}

/// System metrics snapshot
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SystemMetrics {
    pub cpu: f32,
    pub ram: f32,
    pub mic_muted: bool,
    pub audio_playing: bool,
}

/// Telemetry Payload (Desktop -> Phone over Telemetry Characteristic)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TelemetryPayload {
    pub status: String,
    pub active_app: String,
    pub metrics: SystemMetrics,
}

/// Handshake & Pairing Payload (Bidirectional over Auth Characteristic)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct HandshakePayload {
    pub client_id: String,
    pub client_public_key: String,
    pub auth_code: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_action_payload_roundtrip() {
        let action = ActionPayload {
            id: "btn_test_1".to_string(),
            action_type: ActionType::Hotkey,
            modifiers: vec!["PRIMARY_MOD".to_string(), "ALT".to_string()],
            key: "2".to_string(),
            payload: "".to_string(),
            sequence_delay_ms: 0,
        };
        let buf = rmp_serde::to_vec_named(&action).unwrap();
        assert!(buf.len() < 100);
        let decoded: ActionPayload = rmp_serde::from_slice(&buf).unwrap();
        assert_eq!(action, decoded);
    }

    #[test]
    fn test_telemetry_payload_roundtrip() {
        let t = TelemetryPayload {
            status: "READY".to_string(),
            active_app: "com.microsoft.VSCode".to_string(),
            metrics: SystemMetrics { cpu: 14.2, ram: 58.6, mic_muted: false, audio_playing: true },
        };
        let buf = rmp_serde::to_vec_named(&t).unwrap();
        let decoded: TelemetryPayload = rmp_serde::from_slice(&buf).unwrap();
        assert_eq!(t, decoded);
    }
}
